src/io/windows/sync.rs

```diff
@@ -0,0 +1,238 @@
+pub(crate) struct File {
+    inner: Arc<FileInner>
+}
+
+#[derive(Debug)]
+struct FileInner {
+    std_file: StdFile,
+    size: AtomicU64
+}
+
+
+impl File {
+    pub(crate) fn size(&self) -> u64 {
+        self.inner.size.load(Ordering::SeqCst)
+    }
+
+    pub(crate) async fn write_append_writable_data<R: Send + 'static>(
+        &self,
+        c: impl WritableDataCreator<R>,
+    ) -> IOResult<R> {
+        let len = c.len();
+        let file_inner = self.inner.clone();
+        Self::background_sync_call(move || {
+            let offset = file_inner.size.fetch_add(len, Ordering::SeqCst);
+            let (res, data) = c.create(offset);
+            Self::write_data(&file_inner, offset, res)?;
+            Ok(data)
+        })
+        .await
+    }
+
+    fn write_data(file: &FileInner, mut offset: u64, writable_data: WritableData) -> IOResult<()> {
+        match writable_data {
+            WritableData::Single(bytes) => file.write_all_at(&bytes, offset),
+            WritableData::Double(b1, b2) => file.write_all_at(&b1, offset).and_then(|_| {
+                offset = offset + b1.len() as u64;
+                file.write_all_at(&b2, offset)
+            }),
+        }
+    }
+
+    pub(crate) async fn write_append_all(&self, buf: Bytes) -> IOResult<()> {
+        let file_inner = self.inner.clone();
+        Self::background_sync_call(move || {
+            let offset = file_inner.size.fetch_add(buf.len() as u64, Ordering::SeqCst);
+            file_inner.write_all_at(&buf, offset)
+        })
+        .await
+    }
+
+    pub(crate) async fn write_all_at(&self, offset: u64, buf: Bytes) -> IOResult<()> {
+        debug_assert!(offset + buf.len() as u64 <= self.size());
+        let file_inner = self.inner.clone();
+        Self::background_sync_call(move || file_inner.write_all_at(&buf, offset)).await
+    }
+
+    pub(crate) async fn read_all(&self) -> Result<BytesMut> {
+        self.read_exact_at_allocate(self.size().try_into()?, 0).await
+    }
+
+    pub(crate) async fn read_exact_at_allocate(
+        &self,
+        size: usize,
+        offset: u64,
+    ) -> Result<BytesMut> {
+        let buf = BytesMut::zeroed(size);
+        self.read_exact_at(buf, offset).await
+    }
+
+    pub(crate) async fn read_exact_at(&self, mut buf: BytesMut, offset: u64) -> Result<BytesMut> {
+        let file_inner = self.inner.clone();
+        Self::background_sync_call(move || file_inner.read_exact_at(&mut buf, offset).map(|_| buf)).await
+            .map_err(|err| err.into())
+    }
+
+    pub(crate) async fn fsyncdata(&self) -> IOResult<()> {
+        let file_inner = self.inner.clone();
+        Self::background_sync_call(move || file_inner.std_file.sync_all()).await
+    }
+
+    pub(crate) fn created_at(&self) -> IOResult<SystemTime> {
+        let metadata = self.inner.std_file.metadata()?;
+        Ok(metadata.created().unwrap_or(SystemTime::now()))
+    }
+
+    async fn background_sync_call<F, R>(f: F) -> R
```

Comment:
This function and also inplace call can be extracted into separate mod available to all io drivers. Also threshold can be made generic const so each platform will have its own compile-time generated instance of async-sync worker.
This probably can be made into separate ticket

Reply:
I don't think it's needed now. This function is trivial, and this is only its second copy.

I didn't add `inplace_sync_call` to the Windows implementation at all, because the Windows implementation is unlikely to be used in production, and I don't want to spend time running tests to find the best threshold